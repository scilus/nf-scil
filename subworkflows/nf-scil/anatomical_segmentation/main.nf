// ** Importing modules from nf-scil ** //
include { SEGMENTATION_FASTSEG       } from '../../../modules/nf-scil/segmentation/fastseg/main'
include { SEGMENTATION_FREESURFERSEG } from '../../../modules/nf-scil/segmentation/freesurferseg/main'

workflow ANATOMICAL_SEGMENTATION {

    // ** Two input channels for the segmentation processes since they are using   ** //
    // ** different image files. Supplying an empty channel for the one that isn't ** //
    // ** relevant will make the workflow run the appropriate module.              ** //
    take:
        ch_image            // channel: [ val(meta), image ], optional
        ch_freesurferseg    // channel: [ val(meta), aparc_aseg, wmparc ], optional
        ch_lesions          // channel: [ val(meta), lesions ], optional

    main:

        ch_versions = Channel.empty()

        // ** FSL fast segmentation ** //
        // TODO: Add lesion mask
        SEGMENTATION_FASTSEG (
            ch_image
                .join(ch_lesions, remainder: true)
                .map{ it[0..1] + [it[2] ?: []] }
        )
        ch_versions = ch_versions.mix(SEGMENTATION_FASTSEG.out.versions.first())

        // ** Setting outputs ** //
        wm_mask = SEGMENTATION_FASTSEG.out.wm_mask
        gm_mask = SEGMENTATION_FASTSEG.out.gm_mask
        csf_mask = SEGMENTATION_FASTSEG.out.csf_mask
        wm_map = SEGMENTATION_FASTSEG.out.wm_map
        gm_map = SEGMENTATION_FASTSEG.out.gm_map
        csf_map = SEGMENTATION_FASTSEG.out.csf_map

        // ** Freesurfer segmentation ** //
        SEGMENTATION_FREESURFERSEG (
            ch_freesurferseg
                .join(ch_lesions, remainder: true)
                .map{ it[0..2] + [it[3] ?: []] }
        )
        ch_versions = ch_versions.mix(SEGMENTATION_FREESURFERSEG.out.versions.first())

        // ** Setting outputs ** //
        fs_wm_mask = SEGMENTATION_FREESURFERSEG.out.wm_mask
        fs_gm_mask = SEGMENTATION_FREESURFERSEG.out.gm_mask
        fs_csf_mask = SEGMENTATION_FREESURFERSEG.out.csf_mask
        wm_map = Channel.empty()
        gm_map = Channel.empty()
        csf_map = Channel.empty()

    emit:
        wm_mask      = wm_mask                     // channel: [ val(meta), [ wm_mask ] ]
        gm_mask      = gm_mask                     // channel: [ val(meta), [ gm_mask ] ]
        csf_mask     = csf_mask                    // channel: [ val(meta), [ csf_mask ] ]
        wm_map       = wm_map                      // channel: [ val(meta), [ wm_map ] ]
        gm_map       = gm_map                      // channel: [ val(meta), [ gm_map ] ]
        csf_map      = csf_map                     // channel: [ val(meta), [ csf_map ] ]
        fs_wm_mask   = fs_wm_mask                  // channel: [ val(meta), [ wm_mask ] ]
        fs_gm_mask   = fs_gm_mask                  // channel: [ val(meta), [ gm_mask ] ]
        fs_csf_mask  = fs_csf_mask                 // channel: [ val(meta), [ csf_mask ] ]

        versions = ch_versions                     // channel: [ versions.yml ]
}
