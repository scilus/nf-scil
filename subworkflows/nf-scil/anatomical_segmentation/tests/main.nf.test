nextflow_workflow {

    name "Test Subworkflow ANATOMICAL_SEGMENTATION"
    script "../main.nf"
    workflow "ANATOMICAL_SEGMENTATION"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/anatomical_segmentation"

    tag "segmentation"
    tag "segmentation/fastseg"
    tag "segmentation/freesurferseg"

    tag "load_test_data"

        setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "heavy.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("anatomical_segmentation - fslfast") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat/anat_image.nii.gz", checkIfExists: true),
                        []
                    ]}
                input[1] = Channel.empty()
                input[2] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("anatomical_segmentation - freesurferseg") {

        when {
            workflow {
                """
                input[0] = Channel.empty()
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/freesurfer/aparc_aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/freesurfer/wmparc.nii.gz", checkIfExists: true)
                    ]}
                input[2] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("anatomical_segmentation - both") {

        when {
            workflow {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/anat/anat_image.nii.gz", checkIfExists: true)
                    ]}
                input[1] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/freesurfer/aparc_aseg.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/freesurfer/wmparc.nii.gz", checkIfExists: true),
                        []
                    ]}
                input[2] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
