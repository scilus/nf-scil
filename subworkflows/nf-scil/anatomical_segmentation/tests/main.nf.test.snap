{
    "anatomical_segmentation - freesurferseg": {
        "content": [
            {
                "0": [

                ],
                "1": [

                ],
                "2": [

                ],
                "3": [

                ],
                "4": [

                ],
                "5": [

                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,2dc635526adb6ef185300acc326ead13"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,247106ed727133078fe262b0b8244071"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,b92db7e454c95afadea320a143912ce8"
                    ]
                ],
                "9": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "csf_map": [

                ],
                "csf_mask": [

                ],
                "fs_csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,b92db7e454c95afadea320a143912ce8"
                    ]
                ],
                "fs_gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,247106ed727133078fe262b0b8244071"
                    ]
                ],
                "fs_wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,2dc635526adb6ef185300acc326ead13"
                    ]
                ],
                "gm_map": [

                ],
                "gm_mask": [

                ],
                "versions": [
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "wm_map": [

                ],
                "wm_mask": [

                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-02T17:33:19.865748"
    },
    "anatomical_segmentation - fslfast": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,f849e268007a747714e342162f13f631"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,39396421c6d1d32ea7943a5ec7a49dce"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,d0fbc088059f17faff38fbcd33257d27"
                    ]
                ],
                "3": [

                ],
                "4": [

                ],
                "5": [

                ],
                "6": [

                ],
                "7": [

                ],
                "8": [

                ],
                "9": [
                    "versions.yml:md5,0d100de62a24580c4abbed9193fb6352"
                ],
                "csf_map": [

                ],
                "csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,d0fbc088059f17faff38fbcd33257d27"
                    ]
                ],
                "fs_csf_mask": [

                ],
                "fs_gm_mask": [

                ],
                "fs_wm_mask": [

                ],
                "gm_map": [

                ],
                "gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,39396421c6d1d32ea7943a5ec7a49dce"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,f098240b7d7f6df9a2a43215c5828669"
                ],
                "wm_map": [

                ],
                "wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,f849e268007a747714e342162f13f631"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-02T17:27:45.817175"
    },
    "anatomical_segmentation - both": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,f849e268007a747714e342162f13f631"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,39396421c6d1d32ea7943a5ec7a49dce"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,d0fbc088059f17faff38fbcd33257d27"
                    ]
                ],
                "3": [

                ],
                "4": [

                ],
                "5": [

                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,2dc635526adb6ef185300acc326ead13"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,247106ed727133078fe262b0b8244071"
                    ]
                ],
                "8": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,b92db7e454c95afadea320a143912ce8"
                    ]
                ],
                "9": [
                    "versions.yml:md5,0d100de62a24580c4abbed9193fb6352",
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "csf_map": [

                ],
                "csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,d0fbc088059f17faff38fbcd33257d27"
                    ]
                ],
                "fs_csf_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_csf.nii.gz:md5,b92db7e454c95afadea320a143912ce8"
                    ]
                ],
                "fs_gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,247106ed727133078fe262b0b8244071"
                    ]
                ],
                "fs_wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,2dc635526adb6ef185300acc326ead13"
                    ]
                ],
                "gm_map": [

                ],
                "gm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_gm.nii.gz:md5,39396421c6d1d32ea7943a5ec7a49dce"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0d100de62a24580c4abbed9193fb6352",
                    "versions.yml:md5,18dfcb72144006d4f3e2afc759d2ee6b"
                ],
                "wm_map": [

                ],
                "wm_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__mask_wm.nii.gz:md5,f849e268007a747714e342162f13f631"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-02T17:44:03.066837"
    }
}
